Please fill out the information of your group!

| <p style="text-align: center;">First Name</p>  | <p style="text-align: center;">Family Name</p> | Matr.-No. |
| ---------------------------------------------- | ---------------------------------------------- | -------- |
| <p style="text-align: left">Branko</p>        | <p style="text-align: left">Paunović</p>       | <p>k12046370</p> |
| <p style="text-align: left">Vladyslav</p>     | <p style="text-align: left">Savchenko</p>      | <p>k12227285</p> |

<h2 style="text-align: center">344.105/6/7 UE: Natural Language Processing (WS2023/24)</h2>
<h1 style="color:rgb(0,120,170)">Assignment 2</h1>
<h2 style="color:rgb(0,120,170)">Getting to Know Word Embedding!</h2>

<b>Terms of Use</b><br>
This  material is prepared for educational purposes at the Johannes Kepler University (JKU) Linz, and is exclusively provided to the registered students of the mentioned course at JKU. It is strictly forbidden to distribute the current file, the contents of the assignment, and its solution. The use or reproduction of this manuscript is only allowed for educational purposes in non-profit organizations, while in this case, the explicit prior acceptance of the author(s) is required.

**Authors:** Navid Rekab-saz, Oleg Lesota<br>


<h2>Table of contents</h2>
<ol>
    <a href="#section-general-guidelines"><li style="font-size:large;font-weight:bold">General Guidelines</li></a>
    <a href="#section-taskA"><li style="font-size:large;font-weight:bold">Task A: Words Similarity and Nearest Neighbors (15 points)</li></a>
    <a href="#section-taskB"><li style="font-size:large;font-weight:bold">Task B: Document Classification with Word Embedding (15 points)</li></a>
    <a href="#section-taskC"><li style="font-size:large;font-weight:bold">Task C: Classification with sent2vec Document Embeddings (2 extra point)</li></a>
    <a href="#section-references"><li style="font-size:large;font-weight:bold">References</li></a>
    
</ol>

<a name="section-general-guidelines"></a><h2 style="color:rgb(0,120,170)">General Guidelines</h2>

### Assignment objective
The aim of this assignment is to get familiarized with using word embedding (WE) models in practice. The assignment in total has **30 points**; it also offers **2 extra points** which can cover any missing point.

This Notebook encompasses all aspects of the assignment, namely the descriptions of tasks as well as your solutions and reports. Feel free to add any required cell for solutions. The cells can contain code, reports, charts, tables, or any other material, required for the assignment. Feel free to provide the solutions in an interactive and visual way! 

Please discuss any unclear point in the assignment in the provided forum in MOODLE. It is also encouraged to provide answers to your peer's questions. However when submitting a post, keep in mind to avoid providing solutions. Please let the tutor(s) know shall you find any error or unclarity in the assignment.


### Libraries & Dataset

The assignment should be implemented with recent versions of `Python` (>3.7). Any standard Python library can be used, so far that the library is free and can be simply installed using `pip` or `conda`. Examples of potentially useful libraries are `scikit-learn`, `numpy`, `scipy`, `gensim`, `nltk`, `spaCy`, and `AllenNLP`. Use the latest stable version of each library.

To conduct the experiments, we use a subset of the `HumSet` dataset [1] (https://blog.thedeep.io/humset/). `HumSet` is created by the DEEP (https://www.thedeep.io) project – an open source platform which aims to facilitate processing of textual data for international humanitarian response organizations. The platform enables the classification of text excerpts, extracted from news and reports into a set of domain specific classes. The provided dataset contains the classes (labels) referring to the humanitarian sectors like agriculture, health, and protection. The dataset contains an overall number of 17,301 data points. 

Download the dataset from the Moodle page of the course.

the provided zip file consists of the following files:
- `thedeep.subset.train.txt`: Train set in csv format with three fields: sentence_id, text, and label.
- `thedeep.subset.validation.txt`: Validation set in csv format with three fields: sentence_id, text, and label.
- `thedeep.subset.test.txt`: Test set in csv format with three fields: sentence_id, text, and label.
- `thedeep.subset.label.txt`: Captions of the labels.
- `thedeep.ToU.txt`: Terms of use of the dataset.

[1] HumSet: Dataset of Multilingual Information Extraction and Classification for Humanitarian Crises Response
*Selim Fekih, Nicolo' Tamagnone, Benjamin Minixhofer, Ranjan Shrestha, Ximena Contla, Ewan Oglethorpe and Navid Rekabsaz.* 
In Findings of the 2022 Conference on Empirical Methods in Natural Language Processing (Findings of EMNLP), December 2022.



### Submission

Each group should submit the following two files:

- One Jupyter Notebook file (`.ipynb`), containing all the code, results, visualizations, etc. **In the submitted Notebook, all the results and visualizations should already be present, and can be observed simply by loading the Notebook in a browser.** The Notebook must be self-contained, meaning that (if necessary) one can run all the cells from top to bottom without any error. Do not forget to put in your names and student numbers in the first cell of the Notebook. 
- The HTML file (`.html`) achieved from exporting the Jupyter Notebook to HTML (Download As HTML).

You do not need to include the data files in the submission.


<a name="section-taskA"></a><h2 style="color:rgb(0,120,170)">Task A: Words Similarity and Nearest Neighbors (15 points)</h2>


**Loading a word embedding (WE) model (1 points).** Download a pre-trained word embedding model such as word2vec (https://code.google.com/archive/p/word2vec/) or GloVe (https://nlp.stanford.edu/projects/glove/). You can load the downloaded vectors into arrays, or use libraries such as `gensim` to download and process the vectors. 

**Calculating word-to-word similarities (4 points).** Select <ins>5 arbitrary words</ins> from 5 different topics like objects, science disciplines, verbs, adjectives, animals, etc. Let us refer to these words as *source words*. For each source word, calculate its cosine similarities to <ins>6 target words</ins>. The target words of each source word are also selected by you and should cover various levels of semantic relations – according to your linguistic judgement – to the source word, namely from highly-related to not related at all. Organize the target words in tables, such that the target words of each source word are sorted from the highest to the lowest relevance (according to your judgement). Consider the following points:

- **Implementation (2/4 points):** Implement cosine similarity as a function that takes two vectors and returns the similarity score. Implement cosine by yourself and do NOT use the provided functionalities of any library.
- **Reporting and observations (2/4 points):** Report the calculated similarities side by side with your word-to-word semantic relevance judgements in tables. Compare the results and report your observations.  

**Calculating nearest neighbors (10 points).** For the 5 source words, retrieve the $k=10$ nearest neighbors using the word embedding model, namely the words with the highest similarities to the source word. Consider the following points: 
    
- **Overall implementation (3/10 points):** your implemented function takes a source vector, a set of target vectors, and the $k$ parameter, and returns the $k$ nearest neighbors and their similarity scores. Implement nearest neighbor calculation by yourself and do NOT use the provided functionalities of any library.
- **Similarity metrics (2/10 points):** execute the calculation of nearest neighbors according to <ins>two similarity metrics</ins> namely cosine and dot product.
- **Efficiency (3/10 points):** your nearest neighbor functions should provide an *efficient* calculation of nearest neighbors. An inefficient way (which should be avoided!) would be looping over the set of vectors in the word embedding model, and one by one calculating the cosine/dot product similarity of the source vector to each of the target vectors. As a hint for an efficient way, consider that in `numpy` (and other libraries), calculating the dot product of a vector to a matrix is much faster than the dot products of the vector to each vector of the matrix.
- **Reporting and observations (2/10 points):** report the results in tables, which enable comparing between the outputs of the two similarity metrics. Which similarity metric would you prefer? Report your observations.

</div>

In [1]:
import numpy as np
import math

# Loading a word embedding model

In [2]:
def load_glove_model(glove_file):
    """
    :param glove_file: embeddings_path: path of glove file.
    :return: glove model
    """

    f = open(glove_file + ".txt", 'r', encoding='utf-8')
    model = {}
    for line in f:
        split_line = line.split()
        word = split_line[0]
        embedding = np.array([float(val) for val in split_line[1:]])
        model[word] = embedding
    f.close()
    return model

model = load_glove_model("glove.6B.100d")
# here we have 400k words as embedings, pretrained with GloVe model

# Calculating word-to-word similarities

Now choose some words, like: cat, triangle, smart, blue and flying<br>
As tagets words use:<br>
For cat: animal, dog, tail, domesticated, routine, diverges <br>
For triangle: polygon, geometry, pyramids", sharp, egypt, penta <br>
For smart: clever, phone, pet, scientist, gum, anarchy <br>
For blue: color, red, ocean, france, sadness, delaying <br>
For flying: plane, sky, star, leaf, penguin, recursive <br>

In [3]:
def compute_cos(x, y):
    """
    Compute cos(x, y) where x, y are vectors (embeddings)
    :param x: first vector
    :param y: second vector
    :return: distance between given x and y
    """
    xy_dot_prod = 0
    for i in range(x.size):
        xy_dot_prod += x[i]*y[i]
    x_two_norm = np.linalg.norm(x)
    y_two_norm = np.linalg.norm(y)
    return xy_dot_prod / (x_two_norm * y_two_norm)

def display_word_target_cos(word, targets):
    """
    Display similarity between two words:
    :param word: given word, the targets are compared to
    :targets: array of target words
    :return: -
    """
    for target in targets:
        print(f"{word} and {target} similarity score:", compute_cos(model.get(word), model.get(target)))

In [4]:
print("------------------------------------------------------------------------------------------------------")
cat_targets = ["animal", "dog", "tail", "domesticated", "routine", "diverges"]
display_word_target_cos("cat", cat_targets)
print("------------------------------------------------------------------------------------------------------")
triangle_targets = ["polygon", "geometry", "pyramids", "sharp", "egypt", "penta"]
display_word_target_cos("triangle", triangle_targets)
print("------------------------------------------------------------------------------------------------------")
smart_targets = ["clever", "phone", "pet", "scientist", "gum", "anarchy"]
display_word_target_cos("smart", smart_targets)
print("------------------------------------------------------------------------------------------------------")
blue_targets = ["color", "red", "ocean", "france", "sadness", "delaying"]
display_word_target_cos("blue", blue_targets)
print("------------------------------------------------------------------------------------------------------")
flying_targets = ["plane", "sky", "star", "leaf", "penguin", "recursive"]
display_word_target_cos("flying", flying_targets)
print("------------------------------------------------------------------------------------------------------")

------------------------------------------------------------------------------------------------------
cat and animal similarity score: 0.6200780789843513
cat and dog similarity score: 0.8798075041293155
cat and tail similarity score: 0.498913568097558
cat and domesticated similarity score: 0.37443498622822186
cat and routine similarity score: 0.282175848392274
cat and diverges similarity score: 0.0045659344322218675
------------------------------------------------------------------------------------------------------
triangle and polygon similarity score: 0.4187617562438889
triangle and geometry similarity score: 0.35469458466248005
triangle and pyramids similarity score: 0.2920967194861111
triangle and sharp similarity score: 0.2135364739664046
triangle and egypt similarity score: 0.18267866148186557
triangle and penta similarity score: -0.14578928331081956
------------------------------------------------------------------------------------------------------
smart and clever similari

## Reporting and observations
As expected, the more semantically "closer" target words, the more score they achieve by cosine function. Almost all semantical predictions are ranked by cos as expected. However we made a small mistake, when general catagories were ranked above "similar" words. E.g.:blue and red have more semantic relations in comarison to blue and color. But generally semantical similarity has reasonable corelation the score. <br>
The tables for "blue" and "cat", since other words have the same ranking: <br>
color", "red", "ocean", "france", "sadness", "delaying
#### For "cat":
| rank | Manually constructed  | By cos score |
| :- | -: | :-: |
| 1 | animal | dog
| 2 | dog | animal 
| 3 | tail | tail
| 4 | domesticated | domesticated
| 5 | routine | routine
| 6 | diverges | diverges
#### For "blue":
| rank | Manually constructed  | By cos score |
| :- | -: | :-: |
| 1 | color | red
| 2 | red | color 
| 3 | ocean | ocean
| 4 | france | france
| 5 | sadness | sadness
| 6 | delaying | delaying

## Calculating nearest neighbors

In [5]:
# to be honest I think vect_length shold be hardcoded, but anyway let it be like this for now 
vect_length = model.get(list(model.keys())[0]).size

id_word = {}
# matrix, where each row is represented by word and 
words_matrix = np.zeros((len(model),vect_length))
for i, word in enumerate(model):
    id_word[i] = word
    words_matrix[i] = model.get(word)
print(words_matrix.shape)

(400000, 100)


In [6]:
def efficient_top_k_nearest_neighbors_cos(word, words_matrix, model, k):
    """
    Finds top 10 words with highest score with cos function
    :param word: the word, k nearest neighbors shuld be found
    :param words_matrix: matrix of words x embeddings 
    :param model: model to map the word 
    :param model: model to map the word 
    :return: list of tuples (word, score), where word with the greatest score goes first and score decreases over iterations
    """
    word_embedding = model.get(word)
    # single value 2 norm of word 
    x_euclidean_norm = np.linalg.norm(word_embedding)
    # vector of length equal to length of dictionary of model with 2 norms
    y_euclidean_norms = np.linalg.norm(words_matrix, axis = 1)
    # substitute to cos function
    result_matrix = words_matrix.dot(word_embedding) / (x_euclidean_norm * y_euclidean_norms)
    # we doesn't need last elem, since it is word itself
    top_indices = np.argsort(result_matrix)[-(k+1):-1][::-1]
    result = []
    for i, ind in enumerate(top_indices):
        result.append((id_word.get(ind), result_matrix[ind]))
    return result

def efficient_top_k_nearest_neighbors_dot_prod(word, words_matrix, model, k):
    """
    Finds top 10 words with highest score with dot product only
    :param word: the word, k nearest neighbors shuld be found
    :param words_matrix: matrix of words x embeddings 
    :param model: model to map the word 
    :param model: model to map the word 
    :return: list of tuples (word, score), where word with the greatest score goes first and score decreases over iterations
    """
    word_embedding = model.get(word)
    # single value 2 norm of word 
    result_matrix = words_matrix.dot(word_embedding) 
    # we doesn't need last elem, since it is word itself
    top_indices = np.argsort(result_matrix)[-(k+1):-1][::-1]
    result = []
    for i, ind in enumerate(top_indices):
        result.append((id_word.get(ind), result_matrix[ind]))
    return result

k = 10
words = ["cat", "triangle", "smart", "blue", "flying"]
for word in words:
    print("WORD: ", word, "-----------------------------------------------------------------------------")
    top_k_cos = efficient_top_k_nearest_neighbors_cos(word, words_matrix, model, k)
    top_k_dotprod = efficient_top_k_nearest_neighbors_dot_prod(word, words_matrix, model, k)
    print("\nUsing cos function:")
    for word_in_k, score in top_k_cos: 
        print(f"{word} and {word_in_k} has similarity: {score:.4f}")
    print("\nUsing dot product:")
    for word_in_k, score in top_k_dotprod: 
        print(f"{word} and {word_in_k} has similarity: {score:.4f}")

WORD:  cat -----------------------------------------------------------------------------

Using cos function:
cat and dog has similarity: 0.8798
cat and rabbit has similarity: 0.7424
cat and cats has similarity: 0.7323
cat and monkey has similarity: 0.7289
cat and pet has similarity: 0.7190
cat and dogs has similarity: 0.7164
cat and mouse has similarity: 0.6915
cat and puppy has similarity: 0.6800
cat and rat has similarity: 0.6641
cat and spider has similarity: 0.6501

Using dot product:
cat and dog has similarity: 24.9974
cat and dogs has similarity: 20.3302
cat and pet has similarity: 19.7467
cat and mouse has similarity: 19.4296
cat and rabbit has similarity: 19.1802
cat and monkey has similarity: 18.8589
cat and cats has similarity: 18.6155
cat and bird has similarity: 18.5562
cat and baby has similarity: 18.3632
cat and horse has similarity: 18.1018
WORD:  triangle -----------------------------------------------------------------------------

Using cos function:
triangle and equ

## Reporting and observations
### Draw comparison tables for easier readability
#### For cat:
| rank | cos ranked | cos score | dot product ranked | dot product score 
| --- | --- | --- | --- | ---
| 1 | dog | 0.8798 | dog | 24.9974
| 2 | rabbit | 0.7424 | dogs | 20.3302
| 3 | cats | 0.7323 | pet | 19.7467
| 4 | monkey | 0.7289 | mouse | 19.4296
| 5 | pet | 0.7190 | rabbit | 19.1802
| 6 | dogs | 0.7164 |  monkey | 18.8589
| 7 | mouse | 0.6915 | cats | 18.6155
| 8 | puppy | 0.6800 | bird | 18.5562
| 9 | rat | 0.6641 | baby | 18.3632
| 10 | spider | 0.6501 | horse | 18.1018

#### For triangle:
| rank | cos ranked | cos score | dot product ranked | dot product score 
| --- | --- | --- | --- | ---
| 1 | equilateral | 0.5825 | equilateral | 16.8406
| 2 | loop | 0.5716 | loop | 15.8079
| 3 | circle | 0.5570 | axis | 15.6629
| 4 | triangles | 0.5476 | triangular | 15.0958
| 5 | ring | 0.5245 | triangles | 14.9625
| 6 | axis | 0.5126 |  vertices | 14.7652
| 7 | triangular | 0.5064 | angle | 14.5943
| 8 | hypotenuse | 0.5061 | ring | 14.3210
| 9 | angle | 0.5053 | junction | 14.3126
| 10 | outer | 0.5036 | outer | 14.3035

#### For smart:
| rank | cos ranked | cos score | dot product ranked | dot product score 
| --- | --- | --- | --- | ---
| 1 | intelligent | 0.7553 | intelligent | 19.9191
| 2 | sophisticated | 0.6903 | you | 19.4773
| 3 | clever | 0.6687 | sophisticated | 19.2363
| 4 | pretty | 0.6568 | really | 19.0290
| 5 | kind | 0.6472 | i | 18.9315
| 6 | really | 0.6461 |  funny | 18.7606
| 7 | incredibly | 0.6443 | 're | 18.5379
| 8 | savvy | 0.6280 | 'm | 18.2884
| 9 | better | 0.6279 | pretty | 18.2542
| 10 | good | 0.6261 | good | 18.1636

#### For blue:
| rank | cos ranked | cos score | dot product ranked | dot product score 
| --- | --- | --- | --- | ---
| 1 | red | 0.8435 | red | 29.5000
| 2 | black | 0.8066 | black | 28.4665
| 3 | pink | 0.7868 | white | 27.1514
| 4 | green | 0.7868 | green | 26.8947
| 5 | purple |  0.7847 | yellow | 26.7866
| 6 | yellow | 0.7777 |  pink | 26.1560
| 7 | gray | 0.7732 | dark | 24.8052
| 8 | bright | 0.7431 | purple | 24.5768
| 9 | white | 0.7365 | bright | 23.7360
| 10 | dark | 0.7170 | gray | 23.4145

#### For flying:
| rank | cos ranked | cos score | dot product ranked | dot product score 
| --- | --- | --- | --- | ---
| 1 | flew | 0.7707 | aircraft | 27.7549
| 2 | fly | 0.7651 | plane | 26.2015
| 3 | planes | 0.7625 | planes | 26.0612
| 4 | jet | 0.7616 | flight | 25.5058
| 5 | aircraft |  0.7613 | air | 24.3543
| 6 | plane | 0.7533 |  jet | 24.1254
| 7 | flight | 0.7430 | fly | 23.0670
| 8 | airplane | 0.7300 | helicopter | 22.7373
| 9 | airplanes | 0.7134 | flights | 22.5178
| 10 | air | 0.7094 | helicopters | 21.8078

Top k words ranked by cosine function and dot product, as expected, made a lot of sence. By obeserving the result, we can clearly see, that cosine funcion gives more insightufl information in comparison to dot product. Also normalized values of cosine є [-1; 1] are easier to precieve.   

<a name="section-taskB"></a><h2 style="color:rgb(0,120,170)">Task B: Document Classification with WE (15 points)</h2>


This task follows the same instruction for document classification as provided in Assignment 1. You are indeed free to reuse any part of your code in Assignment 1 for this task. In Assignment 1, the representation of each document was created using a bag of words representation followed by dimensionality reduction. In this task, the document representations are created from the pre-trained word embeddings.

**Map word embeddings to dictionary words (5 points).** For every word in the dictionary (as discussed and created in Assignment 1), fetch the corresponding word embedding from the pre-trained model. If no embedding is found, initialize the corresponding word embedding randomly.

**Document embedding as the average of word embeddings (5 points).** Using the word embeddings, the representation of each document is defined as the *mean of the vectors of each document's words*. In particular, given the document $d$, consisting of words $\left[ v_1, v_2, ..., v_{|d|} \right]$, the document representation $\mathbf{e}_d$ is defined as:

$\mathbf{e}_d = \frac{1}{|d|}\sum_{i=1}^{|d|}{\mathbf{e}_{v_i}}$

where $\mathbf{e}_{v}$ is the vector of the word $v$, and $|d|$ is the length of the document.

**Classification and evaluation (5 points)** Using these new document representations, apply <ins>three classification algorithms</ins> and report the evaluation results (based on accuracy metric) on the test set.



<a name="section-taskC"></a><h2 style="color:rgb(0,120,170)">Task C: Classification with sent2vec Document Embeddings (2 extra point)</h2>


Sent2vec [1] suggests another unsupervised approach to creating document embeddings from the underlying word embeddings. First, using the provided code in the paper, train a sendtvec model on the training set to create document embeddings. Then, repeat Task B while using the document embeddings provided by sent2vec. Similar to Task 2, conduct the classification experiments and report evaluation results.

[1] M. Pagliardini, P. Gupta, and M. Jaggi. Unsupervised Learning of Sentence Embeddings using Compositional n-Gram Features. In Proceedings of the conference of the North American Chapter of the Association for Computational Linguistics (NAACL), 2018.
